dff/utils/testing/telegram.py

```diff
@@ -0,0 +1,182 @@
+            if isinstance(response.reply_markup, telethon.tl.types.ReplyKeyboardHide):
+                if msg.ui is not None:
+                    raise RuntimeError(f"Several messages with ui:\n{msg.ui}\n{types.ReplyKeyboardRemove()}")
+                msg.ui = RemoveKeyboard()
+            if response.geo is not None:
+                location = Location(latitude=response.geo.lat, longitude=response.geo.long)
+                if msg.location is not None:
+                    raise RuntimeError(f"Several messages with location:\n{msg.location}\n{location}")
+                msg.location = location
+        return msg
+
+    @asynccontextmanager
+    async def run_bot(self):
+        """A context manager that starts a bot defined in `pipeline`"""
+        self.pipeline.messenger_interface.timeout = 2
+        self.pipeline.messenger_interface.long_polling_timeout = 2
+        task = asyncio.create_task(self.pipeline.messenger_interface.connect(self.pipeline._run_pipeline))
+        logging.info("Bot created")
+        yield
+        self.pipeline.messenger_interface.stop()
+        await task
+
+    async def send_and_check(self, message: Message, tmp_dir):
+        """Send a message from a bot, receive it as client, verify it."""
+        await self.forget_previous_updates()
+
+        async with self.client:
+            messenger_interface = self.pipeline.messenger_interface
+            assert isinstance(messenger_interface, PollingTelegramInterface)
+
+            messages = await self.client.get_messages(self.bot, limit=1)
+            if len(messages) == 0:
+                last_message_id = 0
+            else:
+                last_message_id = messages[0].id
+
+            messenger_interface.messenger.send_response((await self.client.get_me(input_peer=True)).user_id, message)
+
+            await asyncio.sleep(3)
+            bot_messages = [
+                x async for x in self.client.iter_messages(self.bot, min_id=last_message_id, from_user=self.bot)
+            ]
+            bot_messages.reverse()
+            result = await self.parse_responses(bot_messages, tmp_dir)
+
+            assert result == message
+
+    async def forget_previous_updates(self):
+        async with self.run_bot():
+            await asyncio.sleep(7)
+
+    async def check_happy_path(self, happy_path, tmp_dir):
+        await self.forget_previous_updates()
+
+        async with self.client:
+            async with self.run_bot():
+                bot_messages = []
+                last_message = None
+                for request, response in happy_path:
+                    logging.info("Sending request")
+                    user_message = await self.send_message(TelegramMessage.parse_obj(request), bot_messages)
+                    if user_message is not None:
+                        last_message = user_message
+                    logging.info("Request sent")
+                    await asyncio.sleep(3)
```

Comment:
This is experimentally approved time?

Reply:
Reduced it to two [here](https://github.com/deeppavlov/dialog_flow_framework/pull/13/commits/5c8dc3b434931faf7c1327fcaa21a6a2979911b9).
Cannot go lower -- that time is required so that the client will receive the updates.